### ISSUE TITLE
Samplerate in WebAssembly version

### ISSUE THREAD

**Issue by fquirin** [Picovoice/Porcupine#149]
### Expected behavior
16kHz, 44kHz and 48kHz samplerate all work with the same quality.

### Actual behavior
I'm not sure ^^, but I think 16kHz is required.

### Steps to reproduce the behavior
Use the WebAssembly version on different devices and check the samplerate.

Hi Alireza,

I've been experimenting with the WebAssemply version and implemented it into the SEPIA client to work on browser, Android and iOS (https://github.com/SEPIA-Framework/sepia-html-client-app/tree/dev). Good news is that it works in general :-) The strange thing is that it sometimes requires that the user speaks veeeerrryyy sloooowww to trigger the wake-word ^^ and I was wondering if it might be a performance thing (my Android is only mid-class device) or related to samplerate?

Due to limitations of Javascript the recording samplerate is not really controllable (it is more of a hardware setting), so I've checked the different devices to see what samplerate I get:
Depending on how I setup my microphone in Windows I get either 16kHz or 48kHz and both work pretty well, but I have a feeling the 16kHz returns less false-positives.
In Android I have no control over the microphone so I always get 48kHz samplerate and recognition quality is rather bad. Here it is most noticeable that it works better when I speak very slowly o_O.

Do you think the samplerate could influence the recognition in such a way?

**Comment by kenarsa** [comment]
Super cool that it works. There are two things that can cause it one is that your device is too slow to process in real time and is dropping frames of audio. The second thing is the plumbing around the library is using a simple resampling algorithm. So improving that can potentially help. I am interested to learn more about your findings.

**Comment by fquirin** [comment]
I'll probably test the performance aspect on the weekend with an iPhone XR, that should give some insights :-) In the meantime could you point me to the lines of code where the resampling happens?
When I played around with Sphinx4 in the old days (^^) resampling usually was an absolute no-go, do you have any data on how this affects accuracy in Porcupine?

**Comment by kenarsa** [comment]
Cool! Upsampling is a no go (8K -> 16K) as the content is lost already. Downsampling (48 -> 16k etc.) is OK as long as done right.

Here is the portion doing resampling: https://github.com/Picovoice/Porcupine/blob/master/demo/js/picovoiceAudioManager.js#L22-L39

If you find a better way (which I am sure you can, given time) please merge back :)

Also, you can do async processing using Worklets https://developer.mozilla.org/en-US/docs/Web/API/Worklet#Browser_compatibility if you end up using them please merge back as well! Async can help on compute constraints platforms ...

**Comment by fquirin** [comment]
So it took a bit longer as planned ^^ but I finally tested "Hey SEPIA" on the iPhone XR. Performance was pretty amazing. Speaking very slowly still triggered the keyword but speaking it normally worked equally well and quite accurate. If it's due to the better microphone or the performance ... I don't know. 
The only issue that came back (I thought I had fixed it) is that Porcupine and the speech recognition (open-source or Apple native) don't like each other very much. Since both have to share one audio context I'm trying to shutdown Porcupine, free the audio context and then start the ASR and vice versa but once the ASR was active Porcupine won't work anymore until you restart the app. After the next release version of the app I'll try to fix this again :-)
